cmp/compare.go

```diff
@@ -361,26 +352,55 @@ func (s *state) tryMethod(vx, vy reflect.Value, t reflect.Type) bool {
 		return false
 	}
 
-	eq := s.callFunc(m.Func, vx, vy)
+	eq := s.callTTBFunc(m.Func, vx, vy)
 	s.report(eq, vx, vy)
 	return true
 }
 
-func (s *state) callFunc(f, x, y reflect.Value) bool {
-	got := f.Call([]reflect.Value{x, y})[0].Bool()
-	if s.dsCheck.curr == s.dsCheck.next {
-		// Swapping the input arguments is sufficient to check that
-		// f is symmetric and deterministic.
-		want := f.Call([]reflect.Value{y, x})[0].Bool()
-		if got != want {
-			fn := getFuncName(f.Pointer())
-			panic(fmt.Sprintf("non-deterministic or non-symmetric function detected: %s", fn))
-		}
-		s.dsCheck.curr = 0
-		s.dsCheck.next++
+func (s *state) callTRFunc(f, v reflect.Value) reflect.Value {
+	if !s.dynChecker.Next() {
+		return f.Call([]reflect.Value{v})[0]
+	}
+
+	// Run the function twice and ensure that we get the same results back.
+	// We run in goroutines so that the race detector (if enabled) can detect
+	// unsafe mutations to the input.
+	c := make(chan reflect.Value)
+	go detectRaces(c, f, v)
+	want := f.Call([]reflect.Value{v})[0]
+	if got := <-c; !value.Equal(got, want) {
```

Comment:
> What's a realistic case where this check on the _output_ would fail?

I could envision a memo cell implemented as something like:

```go
type memo struct {
    ready <-chan struct{}
    v interface{}
}

func (m *memo) Set(f func() interface{}) {
    if m.ready != nil {
        return
    }
    m.ready = make(chan struct{})
    go func() {
        m.v = f()
        close(m.ready)
    }()
}

func (m *memo) Get() interface{} {
    <-m.ready
    return m.v
}
```

...where application code considers two memo cells to be equal if their `Get` methods return equal values.

For such a memo cell, you could envision the Transformer accepting a pointer to a larger struct and returning a pointer to a field of type `memo` within it. Calling `value.Equal` on such a memo cell would produce a data race instead of a correct comparison.

(I can imagine a similar scenario occurring with a `*sync.Map`: for example, `(encoding/json).typeEncoder` uses a similar trick to handle recursive types, but it happens to block until the final encoder for the type has been set so it likely would not fail in this case.)